File: 2024/day-06/day6.py
from enum import Enum
import time

def process_input():
  with open("./day-06/input.txt", "r") as f:
    matrix = f.readlines()
    matrix = [list(line.strip()) for line in matrix]
    no_of_rows = len(matrix)
    no_of_cols = len(matrix[0])

    # Get coordinates of guard and obstacles
    guard_coord = (0, 0)
    obstacle_set = set()
    for x in range(no_of_cols):
      for y in range(no_of_rows):
        if matrix[y][x] == "#":
          obstacle_set.add((x, y))
        if matrix[y][x] == "^":
          guard_coord = (x, y)
    
    return matrix, obstacle_set, guard_coord
  
def has_exited(guard_coord):
  no_of_rows = len(matrix)
  no_of_cols = len(matrix[0])
  x, y = guard_coord
  return x < 0 or x >= no_of_rows or y < 0 or y >= no_of_cols

def met_obstacle(guard_coord, current_direction, obstacle_set):
  x, y = guard_coord
  if current_direction == Direction.UP:
    y -= 1
  elif current_direction == Direction.DOWN:
    y += 1
  elif current_direction == Direction.LEFT:
    x -= 1
  elif current_direction == Direction.RIGHT:
    x += 1
  return (x, y) in obstacle_set

def solve_part1():
  x, y = initial_guard_coord
  direction = Direction.UP
  visited = set()

  while True:
    if has_exited((x, y)):
      return visited, len(visited)

    visited.add((x,y))
    
    if met_obstacle((x, y), direction, obstacle_set):
      next_enum_value = (direction.value % len(Direction)) + 1
      direction = Direction(next_enum_value)
      continue

    if direction == Direction.UP:
      y -= 1
    elif direction == Direction.DOWN:
      y += 1
    elif direction == Direction.LEFT:
      x -= 1
    elif direction == Direction.RIGHT:
      x += 1


def has_loop(obstacle_set):
  visited = set()
  direction = Direction.UP
  x, y = initial_guard_coord

  while True:
    if has_exited((x, y)):
      break

    if (x, y, direction) in visited:
      return True
    
    visited.add((x, y, direction))

    if met_obstacle((x, y), direction, obstacle_set):
      direction = Direction(direction.value % len(Direction) + 1)
      continue

    if direction == Direction.UP:
      y -= 1
    elif direction == Direction.DOWN:
      y += 1
    elif direction == Direction.LEFT:
      x -= 1
    elif direction == Direction.RIGHT:
      x += 1

  return False

def solve_part2():
  # Brute force solution is to try every possibility for the new obstacles and check if 
  # the guard is stuck in a loop (visited set has an entry with the same coordinates and direction).
  #
  # Time complexity: O(mn) * O(mn) = O(m^2 * n^2) where matrix is m x n.
  # O(mn) possiblities of new obstacle
  # O(mn) for each simulation
  #
  # We only need to check whatever has been visited before in part 1 as only those coordinates 
  # can be reached by the guard.
  total_loops = 0
  possible_obstacles = visited_in_part1

  for obstacle in possible_obstacles:
    obstacle_set.add(obstacle) # Temporarily add a new obstacle to our set
    total_loops += 1 if has_loop(obstacle_set) else 0
    obstacle_set.remove(obstacle)

  return total_loops

if __name__ == "__main__":
  Direction = Enum("Direction", [("UP", 1), ("RIGHT", 2), ("DOWN", 3), ("LEFT", 4)])
  matrix, obstacle_set, initial_guard_coord = process_input()

  part1_start_time = time.time()
  visited_in_part1, part1_solution = solve_part1()
  part1_duration = time.time() - part1_start_time
  print(f"Part 1: {part1_solution}")
  print(f"Part 1 took {part1_duration:.4f} seconds to complete.")

  part2_start_time = time.time()
  part2_solution = solve_part2()
  part2_duration = time.time() - part2_start_time
  print(f"Part 2: {part2_solution}")
  print(f"Part 2 took {part2_duration:.4f} seconds to complete.")

File: 2024/day-06/day6.go
package main

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

type State struct {
  X int
  Y int
  DIR string
}

type Coord struct {
  X int
  Y int
}

func processInput() ([][]byte, State) {
  input, err := os.ReadFile("./day-06/input.txt")
  if err != nil {
    fmt.Println("Error reading file: ", err)
    os.Exit(1)
  }

  lines := strings.Split(string(input), "\n")
  var mat [][]byte
  for _, line := range lines {
      mat = append(mat, []byte(line))
  }
  
  var start State;
  for y, row := range mat {
    for x, cell := range row {
      if cell == '^' {
        start = State{X: x, Y: y, DIR: "UP"}
      }
    }
  }
  return mat, start
}

func part1() (map[Coord]bool, int) {
  mat, curr := processInput()
  visited := map[Coord]bool{}

  for true {
    if hasExited(mat, curr) {
      return visited, len(visited)
    }

    visited[Coord{X: curr.X, Y: curr.Y}] = true
    
    if metObstacle(mat, curr) {
      curr = turnRight(curr)
    } else {
      curr = move(curr)
    }
  }
  
  return visited, -1
}

func turnRight(curr State) State {
  turns := map[string]string{
    "UP": "RIGHT", "RIGHT": "DOWN", "DOWN": "LEFT", "LEFT": "UP",
  }
  curr.DIR = turns[curr.DIR]
  return curr
}

func move(curr State) State {
  if curr.DIR == "UP" {
    curr.Y -= 1
  } else if curr.DIR == "DOWN" {
    curr.Y += 1
  } else if curr.DIR == "LEFT" {
    curr.X -= 1
  } else if curr.DIR == "RIGHT" {
    curr.X += 1
  }
  return curr
}

func metObstacle(mat [][]byte, curr State) bool {
  curr = move(curr)
  return !hasExited(mat, curr) && mat[curr.Y][curr.X] == '#'
}

func hasExited(mat [][]byte, curr State) bool {
  rows := len(mat)
  cols := len(mat[0])
  return curr.X < 0 || curr.X >= cols || curr.Y < 0 || curr.Y >= rows
}

func part2(possibleObstacles map[Coord]bool) int {
  // All possible obstacles we can add are the visited coords
  // in part 1.
  mat, curr := processInput()
  loopsFound := 0
  var mu sync.Mutex
  var wg sync.WaitGroup


  for obstacle := range possibleObstacles {
    wg.Add(1)
    go func(obstacle Coord) {
      defer wg.Done()

      matCopy := make([][]byte, len(mat))
      for i := range mat {
        matCopy[i] = append([]byte{}, mat[i]...)
      }

      matCopy[obstacle.Y][obstacle.X] = '#'
      if hasLoop(matCopy, curr) {
        mu.Lock()
        loopsFound++
        mu.Unlock()
      }
    }(obstacle)
  }

  wg.Wait()
  return loopsFound
}

func hasLoop(mat [][]byte, curr State) bool {
  visited := map[State]bool{}

  for true {
    if hasExited(mat, curr) {
      break
    }

    if visited[curr] {
      return true
    }
    visited[curr] = true

    if metObstacle(mat, curr) {
      curr = turnRight(curr) 
    } else {
      curr = move(curr)
    }
  }

  return false
}

func main() {
  start := time.Now()
  part1Visited, part1Sol := part1()
  fmt.Println("Part 1 sol: ", part1Sol)
  fmt.Println("Part 1 time: ", time.Since(start))

  start = time.Now()
  fmt.Println("Part 2 sol: ", part2(part1Visited))
  fmt.Println("Part 2 time: ", time.Since(start))
}
